File: packages/java/src/main/java/io/idrto/def/Def.java
package io.idrto.def;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

public final class Def {
    public static final int MAX_LABEL_LENGTH = 63;
    public static final String IDRTO_HASH_MARKER = "idrto-h1--";
    public static final int HASH_BODY_LENGTH = 50;
    public static final String STRUCTURAL_SEPARATOR = "--";
    public static final String STRUCTURAL_SEPARATOR_ESCAPED = "-2d-2d";

    private static final String BASE36_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz";
    private static final char[] HEX_CHARS = "0123456789abcdef".toCharArray();
    private static final byte[] HEX_DIGITS = new byte[256];

    static {
        Arrays.fill(HEX_DIGITS, (byte) -1);
        for (int i = 0; i < 10; i++) {
            HEX_DIGITS['0' + i] = (byte) i;
        }
        for (int i = 0; i < 6; i++) {
            HEX_DIGITS['a' + i] = (byte) (10 + i);
            HEX_DIGITS['A' + i] = (byte) (10 + i);
        }
    }

    public enum ErrorCode {
        LABEL_TOO_LONG,
        INVALID_ESCAPE,
        INVALID_UTF8,
        INVALID_ENCODING,
        INVALID_LOCATOR,
        NOT_DECODABLE
    }

    public static final class DefException extends Exception {
        private final ErrorCode code;

        public DefException(String message, ErrorCode code) {
            super(message);
            this.code = code;
        }

        public ErrorCode getCode() {
            return code;
        }
    }

    private Def() {}

    private static boolean isLiteralByte(int value) {
        return (value >= 0x61 && value <= 0x7a) || (value >= 0x30 && value <= 0x39);
    }

    private static boolean isHostStart(int value) {
        return isLiteralByte(value);
    }

    private static byte[] canonicalizeToBytes(String input) {
        byte[] raw = input.getBytes(StandardCharsets.UTF_8);
        byte[] bytes = new byte[raw.length];
        for (int i = 0; i < raw.length; i++) {
            byte value = raw[i];
            bytes[i] = (value >= 'A' && value <= 'Z') ? (byte) (value + ('a' - 'A')) : value;
        }
        return bytes;
    }

    private static String encodeBytes(byte[] bytes) {
        StringBuilder out = new StringBuilder(bytes.length * 2);
        for (byte value : bytes) {
            int unsigned = value & 0xff;
            if (isLiteralByte(unsigned)) {
                out.append((char) unsigned);
            } else {
                out.append('-');
                out.append(HEX_CHARS[unsigned >>> 4]);
                out.append(HEX_CHARS[unsigned & 0x0f]);
            }
        }
        return out.toString();
    }

    /** Encode input to a DEF body (§9). */
    public static String encodeBody(String input) {
        return encodeBytes(canonicalizeToBytes(input));
    }

    private static int parseHexPair(int h1, int h2) {
        if (h1 < 0 || h1 >= 256 || h2 < 0 || h2 >= 256) {
            return -1;
        }
        int n1 = HEX_DIGITS[h1];
        int n2 = HEX_DIGITS[h2];
        if (n1 < 0 || n2 < 0) {
            return -1;
        }
        return n1 * 16 + n2;
    }

    /** Decode a DEF body (§10). */
    public static String decodeBody(String body) throws DefException {
        byte[] out = new byte[body.length()];
        int outLen = 0;

        for (int i = 0; i < body.length(); ) {
            char ch = body.charAt(i);
            if (isLiteralByte(ch)) {
                out[outLen++] = (byte) ch;
                i++;
                continue;
            }

            if (ch != '-') {
                throw new DefException("invalid character in encoded input", ErrorCode.INVALID_ESCAPE);
            }
            if (i + 3 > body.length()) {
                throw new DefException("truncated escape sequence", ErrorCode.INVALID_ESCAPE);
            }

            int value = parseHexPair(body.charAt(i + 1), body.charAt(i + 2));
            if (value < 0) {
                throw new DefException("invalid escape sequence", ErrorCode.INVALID_ESCAPE);
            }

            out[outLen++] = (byte) value;
            i += 3;
        }

        try {
            var decoder = StandardCharsets.UTF_8.newDecoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappedCharacter(CodingErrorAction.REPORT);
            return decoder.decode(ByteBuffer.wrap(out, 0, outLen)).toString();
        } catch (CharacterCodingException ex) {
            throw new DefException("invalid utf-8 byte sequence", ErrorCode.INVALID_UTF8);
        }
    }

    private static String[] splitLocator(String locator) throws DefException {
        int sep = locator.indexOf(STRUCTURAL_SEPARATOR);
        if (sep <= 0 || sep + 2 >= locator.length()) {
            throw new DefException("invalid profile locator", ErrorCode.INVALID_LOCATOR);
        }

        String host = locator.substring(0, sep);
        String entity = locator.substring(sep + 2);
        if (host.contains(STRUCTURAL_SEPARATOR) || entity.isEmpty()) {
            throw new DefException("invalid profile locator", ErrorCode.INVALID_LOCATOR);
        }

        byte[] hostBytes = host.getBytes(StandardCharsets.UTF_8);
        if (hostBytes.length == 0 || !isHostStart(hostBytes[0] & 0xff)) {
            throw new DefException("invalid profile host", ErrorCode.INVALID_LOCATOR);
        }

        return new String[] { host, entity };
    }

    private static String base36(byte[] data) {
        BigInteger n = new BigInteger(1, data);
        if (n.signum() == 0) {
            return "0".repeat(HASH_BODY_LENGTH);
        }

        char[] digits = new char[HASH_BODY_LENGTH];
        BigInteger base = BigInteger.valueOf(36);
        for (int i = HASH_BODY_LENGTH - 1; i >= 0; i--) {
            BigInteger[] divRem = n.divideAndRemainder(base);
            digits[i] = BASE36_ALPHABET.charAt(divRem[1].intValue());
            n = divRem[0];
        }
        return new String(digits);
    }

    private static void validateMarker(String marker) throws DefException {
        if (marker.length() < 3
                || marker.length() > 13
                || !marker.endsWith(STRUCTURAL_SEPARATOR)
                || marker.startsWith("xn--")) {
            throw new DefException("invalid provider hash marker", ErrorCode.INVALID_ENCODING);
        }
    }

    private static boolean isBase36Char(int code) {
        return (code >= 0x30 && code <= 0x39) || (code >= 0x61 && code <= 0x7a);
    }

    /** Encode a profile locator with the configured hash marker (§12). */
    public static String encodeProfile(String locator) throws DefException {
        return encodeProfile(locator, IDRTO_HASH_MARKER);
    }

    /** Encode a profile locator with the configured hash marker (§12). */
    public static String encodeProfile(String locator, String marker) throws DefException {
        validateMarker(marker);

        byte[] canonical = canonicalizeToBytes(locator);
        String canonicalText = new String(canonical, StandardCharsets.UTF_8);
        String[] parts = splitLocator(canonicalText);
        String host = parts[0];
        String entity = parts[1];

        String hostBody = encodeBytes(host.getBytes(StandardCharsets.UTF_8));
        String entityBody = encodeBytes(entity.getBytes(StandardCharsets.UTF_8));
        String label = hostBody + STRUCTURAL_SEPARATOR + entityBody;

        if (label.length() <= MAX_LABEL_LENGTH && !label.startsWith("xn--")) {
            return label;
        }

        String hashInput = hostBody + STRUCTURAL_SEPARATOR_ESCAPED + entityBody;
        byte[] digest;
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            digest = md.digest(hashInput.getBytes(StandardCharsets.UTF_8));
        } catch (NoSuchAlgorithmException ex) {
            throw new IllegalStateException("SHA-256 not available", ex);
        }

        String encoded = marker + base36(digest);
        if (encoded.length() > MAX_LABEL_LENGTH) {
            throw new DefException("encoded label exceeds 63 characters", ErrorCode.LABEL_TOO_LONG);
        }
        return encoded;
    }

    /** Decode a profile label with the configured hash marker (§12). */
    public static String decodeProfile(String label) throws DefException {
        return decodeProfile(label, IDRTO_HASH_MARKER);
    }

    /** Decode a profile label with the configured hash marker (§12). */
    public static String decodeProfile(String label, String marker) throws DefException {
        validateMarker(marker);

        if (label.length() > MAX_LABEL_LENGTH) {
            throw new DefException("encoded label exceeds 63 characters", ErrorCode.LABEL_TOO_LONG);
        }

        if (label.startsWith(marker)) {
            String digest = label.substring(marker.length());
            if (digest.length() != HASH_BODY_LENGTH) {
                throw new DefException("invalid profile hash label", ErrorCode.INVALID_ENCODING);
            }
            for (int i = 0; i < digest.length(); i++) {
                if (!isBase36Char(digest.charAt(i))) {
                    throw new DefException("invalid profile hash label", ErrorCode.INVALID_ENCODING);
                }
            }
            throw new DefException("profile hash label is not decodable", ErrorCode.NOT_DECODABLE);
        }

        if (label.startsWith("xn--")) {
            throw new DefException("invalid profile label", ErrorCode.INVALID_ENCODING);
        }

        int sep = label.indexOf(STRUCTURAL_SEPARATOR);
        if (sep <= 0 || sep + 2 > label.length()) {
            throw new DefException("missing profile separator", ErrorCode.INVALID_ENCODING);
        }

        String host = decodeBody(label.substring(0, sep));
        String entity = decodeBody(label.substring(sep + 2));

        if (host.isEmpty() || entity.isEmpty() || host.contains(STRUCTURAL_SEPARATOR)) {
            throw new DefException("invalid decoded profile locator", ErrorCode.INVALID_LOCATOR);
        }

        return host + STRUCTURAL_SEPARATOR + entity;
    }

    /** Encode an idr.to identity locator. */
    public static String encode(String locator) throws DefException {
        return encodeProfile(locator, IDRTO_HASH_MARKER);
    }

    /** Decode an idr.to profile label. */
    public static String decode(String label) throws DefException {
        return decodeProfile(label, IDRTO_HASH_MARKER);
    }
}

File: packages/dotnet/Idrto.DnsEncodedFormat/Def.cs
using System.Numerics;
using System.Security.Cryptography;
using System.Text;

namespace Idrto.DnsEncodedFormat;

public static class Def
{
    public const int MaxLabelLength = 63;
    public const string IdrtoHashMarker = "idrto-h1--";
    public const string IdrtoMarkerHost = "idrto-h1";
    public const string ReservedHostXn = "xn";
    public const int HashBodyLength = 50;
    public const string StructuralSeparator = "--";
    public const string StructuralSeparatorEscaped = "-2d-2d";

    private const string Base36Alphabet = "0123456789abcdefghijklmnopqrstuvwxyz";
    private static ReadOnlySpan<byte> Hex => "0123456789abcdef"u8;
    private static readonly UTF8Encoding Utf8Strict = new(false, true);

    public enum ErrorCode
    {
        LabelTooLong,
        InvalidEscape,
        InvalidUtf8,
        InvalidEncoding,
        InvalidLocator,
        NotDecodable
    }

    public sealed class DefException : Exception
    {
        public DefException(string message, ErrorCode code) : base(message)
        {
            Code = code;
        }

        public ErrorCode Code { get; }
    }

    private static bool IsLiteralByte(byte value) =>
        (value >= 0x61 && value <= 0x7a) || (value >= 0x30 && value <= 0x39);

    private static bool IsHostStart(byte value) => IsLiteralByte(value);

    private static bool IsBase36Char(char c) =>
        (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z');

    private static byte[] CanonicalizeToBytes(string input)
    {
        var bytes = Encoding.UTF8.GetBytes(input);
        for (var i = 0; i < bytes.Length; i++)
        {
            if (bytes[i] is >= 0x41 and <= 0x5a)
            {
                bytes[i] += 32;
            }
        }

        return bytes;
    }

    private static string EncodeBytes(ReadOnlySpan<byte> bytes)
    {
        var builder = new StringBuilder(bytes.Length * 3);
        foreach (var value in bytes)
        {
            if (IsLiteralByte(value))
            {
                builder.Append((char)value);
            }
            else
            {
                builder.Append('-');
                builder.Append((char)Hex[value >> 4]);
                builder.Append((char)Hex[value & 0x0f]);
            }
        }

        return builder.ToString();
    }

    /// <summary>Encode input to a DEF body (§9).</summary>
    public static string EncodeBody(string input) =>
        EncodeBytes(CanonicalizeToBytes(input));

    private static int? ParseHexByte(char h1, char h2)
    {
        int? Hi(char c) => c switch
        {
            >= '0' and <= '9' => c - '0',
            >= 'a' and <= 'f' => c - 'a' + 10,
            _ => null
        };

        var hi = Hi(h1);
        var lo = Hi(h2);
        return hi is null || lo is null ? null : hi.Value * 16 + lo.Value;
    }

    /// <summary>Decode a DEF body (§10).</summary>
    public static string DecodeBody(string body)
    {
        var bytes = new byte[body.Length];
        var len = 0;

        for (var i = 0; i < body.Length;)
        {
            var code = body[i];
            if (IsLiteralByte((byte)code))
            {
                bytes[len++] = (byte)code;
                i++;
                continue;
            }

            if (code != '-')
            {
                throw new DefException("invalid character in encoded input", ErrorCode.InvalidEscape);
            }

            if (i + 3 > body.Length)
            {
                throw new DefException("truncated escape sequence", ErrorCode.InvalidEscape);
            }

            var value = ParseHexByte(body[i + 1], body[i + 2]);
            if (value is null)
            {
                throw new DefException("invalid escape sequence", ErrorCode.InvalidEscape);
            }

            bytes[len++] = (byte)value;
            i += 3;
        }

        try
        {
            return Utf8Strict.GetString(bytes.AsSpan(0, len));
        }
        catch (DecoderFallbackException)
        {
            throw new DefException("invalid utf-8 byte sequence", ErrorCode.InvalidUtf8);
        }
    }

    private static string MarkerHostPrefix(string marker)
    {
        if (!marker.EndsWith(StructuralSeparator, StringComparison.Ordinal))
        {
            throw new DefException("invalid provider hash marker", ErrorCode.InvalidEncoding);
        }

        return marker[..^StructuralSeparator.Length];
    }

    private static void ValidateHost(string host, string marker)
    {
        if (host == ReservedHostXn || host == MarkerHostPrefix(marker))
        {
            throw new DefException("invalid profile host", ErrorCode.InvalidLocator);
        }
    }

    private static (string Host, string Entity) SplitLocator(string locator, string marker)
    {
        var sep = locator.IndexOf(StructuralSeparator, StringComparison.Ordinal);
        if (sep <= 0 || sep + 2 >= locator.Length)
        {
            throw new DefException("invalid profile locator", ErrorCode.InvalidLocator);
        }

        var host = locator[..sep];
        var entity = locator[(sep + 2)..];
        if (host.Contains(StructuralSeparator, StringComparison.Ordinal) || entity.Length == 0)
        {
            throw new DefException("invalid profile locator", ErrorCode.InvalidLocator);
        }

        var hostBytes = Encoding.UTF8.GetBytes(host);
        if (hostBytes.Length == 0 || !IsHostStart(hostBytes[0]))
        {
            throw new DefException("invalid profile host", ErrorCode.InvalidLocator);
        }

        ValidateHost(host, marker);
        return (host, entity);
    }

    private static string Base36(ReadOnlySpan<byte> data)
    {
        var n = new BigInteger(data, isUnsigned: true, isBigEndian: true);
        if (n.IsZero)
        {
            return new string('0', HashBodyLength);
        }

        var builder = new StringBuilder(HashBodyLength);
        while (n > 0)
        {
            n = BigInteger.DivRem(n, 36, out var rem);
            builder.Append(Base36Alphabet[(int)rem]);
        }

        var chars = builder.ToString().ToCharArray();
        Array.Reverse(chars);
        var outValue = new string(chars);
        return outValue.Length < HashBodyLength
            ? new string('0', HashBodyLength - outValue.Length) + outValue
            : outValue;
    }

    private static void ValidateMarker(string marker)
    {
        if (marker.Length < 3
            || marker.Length > 13
            || !marker.EndsWith(StructuralSeparator, StringComparison.Ordinal)
            || marker.StartsWith("xn--", StringComparison.Ordinal))
        {
            throw new DefException("invalid provider hash marker", ErrorCode.InvalidEncoding);
        }
    }

    /// <summary>Encode a profile locator with the configured hash marker (§12).</summary>
    public static string EncodeProfile(string locator, string marker = IdrtoHashMarker)
    {
        ValidateMarker(marker);

        var canonical = CanonicalizeToBytes(locator);
        var canonicalText = Utf8Strict.GetString(canonical);
        var (host, entity) = SplitLocator(canonicalText, marker);

        var hostBody = EncodeBytes(Encoding.UTF8.GetBytes(host));
        var entityBody = EncodeBytes(Encoding.UTF8.GetBytes(entity));
        var label = hostBody + StructuralSeparator + entityBody;

        if (label.Length <= MaxLabelLength)
        {
            return label;
        }

        var hashInput = hostBody + StructuralSeparatorEscaped + entityBody;
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(hashInput));
        var encoded = marker + Base36(digest);
        if (encoded.Length > MaxLabelLength)
        {
            throw new DefException("encoded label exceeds 63 characters", ErrorCode.LabelTooLong);
        }

        return encoded;
    }

    /// <summary>Decode a profile label with the configured hash marker (§12).</summary>
    public static string DecodeProfile(string label, string marker = IdrtoHashMarker)
    {
        ValidateMarker(marker);

        if (label.Length > MaxLabelLength)
        {
            throw new DefException("encoded label exceeds 63 characters", ErrorCode.LabelTooLong);
        }

        if (label.StartsWith(marker, StringComparison.Ordinal))
        {
            var digest = label[marker.Length..];
            if (digest.Length != HashBodyLength)
            {
                throw new DefException("invalid profile hash label", ErrorCode.InvalidEncoding);
            }

            foreach (var c in digest)
            {
                if (!IsBase36Char(c))
                {
                    throw new DefException("invalid profile hash label", ErrorCode.InvalidEncoding);
                }
            }

            throw new DefException("profile hash label is not decodable", ErrorCode.NotDecodable);
        }

        if (label.StartsWith("xn--", StringComparison.Ordinal))
        {
            throw new DefException("invalid profile label", ErrorCode.InvalidEncoding);
        }

        var sep = label.IndexOf(StructuralSeparator, StringComparison.Ordinal);
        if (sep <= 0 || sep + 2 > label.Length)
        {
            throw new DefException("missing profile separator", ErrorCode.InvalidEncoding);
        }

        var host = DecodeBody(label[..sep]);
        var entity = DecodeBody(label[(sep + 2)..]);

        if (host.Length == 0
            || entity.Length == 0
            || host.Contains(StructuralSeparator, StringComparison.Ordinal))
        {
            throw new DefException("invalid decoded profile locator", ErrorCode.InvalidLocator);
        }

        ValidateHost(host, marker);

        return host + StructuralSeparator + entity;
    }

    /// <summary>Encode an idr.to identity locator.</summary>
    public static string Encode(string locator) => EncodeProfile(locator, IdrtoHashMarker);

    /// <summary>Decode an idr.to profile label.</summary>
    public static string Decode(string label) => DecodeProfile(label, IdrtoHashMarker);
}
